### PR TITLE
Refactor/#63 refactor c layer

### PR DESCRIPTION
**Pull request by tomuben**

### CODE REVIEW

File: src/connection.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+SEXP createWriteConnection(pRODBCHandle handle, SQLCHAR *query) {
+    SEXP retVal = nullptr;
+    auto connectionContext = exa::gpConnectionContext;
+    if (connectionContext != nullptr && connectionContext->mConnectionController) {
+        exa::writer::Writer *writer =
+                connectionContext->mConnectionController->startWriting(exa::OdbcSessionInfoImpl(handle, query),
+                                                                       exa::ProtocolType::http);
+        if (writer != nullptr) {
+            auto writeConnection = std::make_unique<exa::rconnection::RWriterConnection>(*writer);
```
[the remark this answers is not in the document]
Comment (tkilias) [reply src/connection.cpp:79]: same for createReadConnection

File: src/r-exasol/impl/connection/ConnectionFactoryImpl.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+std::unique_ptr<exa::reader::Reader> exa::ConnectionFactoryImpl::createHttpReader(Socket& socket) {
+    return std::make_unique<exa::reader::HttpChunkReader>(socket, getChunk());
+}
+
+std::unique_ptr<exa::writer::Writer> exa::ConnectionFactoryImpl::createHttpWriter(Socket& socket) {
+    return std::make_unique<exa::writer::HttpChunkWriter>(socket, getChunk());
+}
+
+exa::Chunk & exa::ConnectionFactoryImpl::getChunk() {
```
Comment (tkilias) [inline src/r-exasol/impl/connection/ConnectionFactoryImpl.cpp:22]: ????

File: src/r-exasol/impl/connection/protocol/http/reader/HttpChunkReader.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        goto error;
+    }
+
+    mChunk.chunk_len = buflen - 2;
+    mChunk.chunk_pos = 0;
+    mChunk.chunk_buf[buflen-2] = '\0';
+    mChunk.chunk_num ++;
+    return mChunk.chunk_len;
+
+    error:
```
Comment (tkilias) [inline src/r-exasol/impl/connection/protocol/http/reader/HttpChunkReader.cpp:77]: I guess, you didn't change this much yet.

**Review by tkilias — COMMENTED** [review]
Preliminary review, see my comment, the remaining files, I need to check another time

File: src/r-exasol/impl/rconnection/RWriterConnection.cpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    mConn->isopen = TRUE;
+    mConn->blocking = TRUE;
+    mConn->canread = FALSE;
+    mConn->canwrite = TRUE;
+    mConn->vfprintf = &pipe_vfprintf;
+    mConn->write = &pipe_write;
+    mConn->fflush = &pipe_fflush;
+    mConn->save = -1000;
+    mConn->priv = (void*)::malloc(sizeof(RWriterConnection*));
+    *(static_cast<writer::Writer**>(mConn->priv)) = &mWriter;
```
Comment (tkilias) [inline src/r-exasol/impl/rconnection/RWriterConnection.cpp:69]: WTF

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    mConn->isopen = TRUE;
+    mConn->blocking = TRUE;
+    mConn->canread = TRUE;
+    mConn->canwrite = FALSE;
+    mConn->read = &pipe_read;
+    mConn->fgetc = &dummy_fgetc;
+    mConn->fgetc_internal = &file_fgetc_internal;
+    mConn->save = -1000;
+    mConn->priv = (void*)::malloc(sizeof(reader::Reader*));
+    *(static_cast<reader::Reader**>(mConn->priv)) = &mReader;
```
Comment (tkilias) [inline src/r-exasol/impl/rconnection/RReaderConnection.cpp:62]: WTF

File: src/connection.cpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+#include <r-exasol/impl/odbc/OdbcSessionInfoImpl.h>
+
+namespace exa {
+    struct ConnectionContext {
+        std::unique_ptr <exa::rconnection::RConnection> mConnection;
+        std::unique_ptr <ConnectionController> mConnectionController;
+    };
+
+    ConnectionFactoryImpl gConnectionFactory;
+    ConnectionContext *gpConnectionContext = nullptr;
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/connection.cpp:20]: Changed to a unique_ptr. 
Why a static variable. Currently we have only one connection, so there is no need to pass the pointer to the client.

File: src/connection.cpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+SEXP createWriteConnection(pRODBCHandle handle, SQLCHAR *query) {
+    SEXP retVal = nullptr;
+    auto connectionContext = exa::gpConnectionContext;
+    if (connectionContext != nullptr && connectionContext->mConnectionController) {
+        exa::writer::Writer *writer =
+                connectionContext->mConnectionController->startWriting(exa::OdbcSessionInfoImpl(handle, query),
+                                                                       exa::ProtocolType::http);
+        if (writer != nullptr) {
+            auto writeConnection = std::make_unique<exa::rconnection::RWriterConnection>(*writer);
+            retVal = writeConnection->create();
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/connection.cpp:83]: Create returns a SEXP (which needs to be protected with PROTECt/UNPROTECT). If I would run create in the constructor I would have to store the SEXP as member, which then would have to be accessed by the client later, this adds a certain complexity.

File: src/r-exasol/if/Writer.h [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+#include <cstdio>
+
+namespace exa {
+    namespace writer {
+        class Writer {
+        public:
+            virtual ~Writer() = default;
+            virtual void start() = 0;
+            virtual size_t pipe_write(const void *ptr, size_t size, size_t nitems) = 0;
+            virtual int pipe_fflush()  = 0;
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/r-exasol/if/Writer.h:17]: Yes, see here https://www.cs.kent.ac.uk/projects/cxxr/doc/html/Rconnections_8h_source.html

File: src/r-exasol/impl/connection/ConnectionController.cpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
+//
+// Created by thomas on 14/07/2021.
+//
+
+#include <r-exasol/impl/connection/ConnectionController.h>
+#include <r-exasol/impl/connection/ConnectionException.h>
+#include <r-exasol/impl/connection/protocol/metaInfoReader.h>
+#include <iostream>
+
+exa::ConnectionController::ConnectionController(ConnectionFactory &connectionFactory, const tErrorFunction & errorHandler)
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/r-exasol/impl/connection/ConnectionController.cpp:11]: The chunk is used among the writer/reader. For performance issues it is allocated as global variable and then just reused. But thus, as forces the factory to be none-const.

File: src/r-exasol/impl/connection/ConnectionFactoryImpl.cpp (hunk 2 above)
Comment (tomuben) [reply src/r-exasol/impl/connection/ConnectionFactoryImpl.cpp:22]: the chunk is implemented as a singleton here.

File: src/r-exasol/impl/connection/protocol/http/reader/HttpChunkReader.cpp (hunk 3 above)
Comment (tomuben) [reply src/r-exasol/impl/connection/protocol/http/reader/HttpChunkReader.cpp:77]: Yeah, as the idea is to eventually use an http library, my thought was to not spend to much effort right now...

File: src/r-exasol/impl/rconnection/RWriterConnection.cpp (hunk 4 above)
Comment (tomuben) [reply src/r-exasol/impl/rconnection/RWriterConnection.cpp:73]: I added a comment to explain what's going on....

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp (hunk 5 above)
Comment (tomuben) [reply src/r-exasol/impl/rconnection/RReaderConnection.cpp:66]: I added a comment to explain what's going on....

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    return  retVal;
+}
+
+}
+
+namespace rcon = exa::rconnection;
+
+rcon::RReaderConnection::RReaderConnection(reader::Reader & reader)
+: mReader(reader)
+, mConn(nullptr){}
```
Comment (tkilias) [inline src/r-exasol/impl/rconnection/RReaderConnection.cpp:47]: mConn is not a pointer, how can you assign a nullptr

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+SEXP rcon::RReaderConnection::create() {
+    SEXP r_custom_connection;
+    PROTECT(r_custom_connection = R_new_custom_connection("exasol", "r", "textConnection", &mConn));
```
Comment (tkilias) [inline src/r-exasol/impl/rconnection/RReaderConnection.cpp:51]: I guess this is the pipe which we use to connect the R world with the http socket. We should document this somewhere.

File: src/r-exasol/impl/rconnection/RWriterConnection.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+        class RWriterConnection : public RConnection {
+        public:
+            explicit RWriterConnection(writer::Writer & writer);
+
+            SEXP create();
+            void release() override;
+
+        private:
+            writer::Writer & mWriter;
+            Rconnection mConn;
```
Comment (tkilias) [inline src/r-exasol/impl/rconnection/RWriterConnection.h:24]: We really need a better distinction between Rconnection (which comes from R I guess and seems something which we can copy) and RConnection which we have defined. Maybe, we could make it more explicit with ExaRConnection or CppRConnection.

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp (hunk 5 above)
Comment (tkilias) [reply src/r-exasol/impl/rconnection/RReaderConnection.cpp:66]: so basically, we would need a shared pointer anything to solve this properly.

File: src/r-exasol/ConnectionContext.cpp [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+#include <r-exasol/odbc/OdbcSessionInfoImpl.h>
+
+namespace exa {
+    void onError(std::string e) {
+        ::error(e.c_str());
+    }
+}
+
+int exa::ConnectionContext::initConnection(const char *host, int port) {
+    destroyConnection(1);
```
Comment (tkilias) [inline src/r-exasol/ConnectionContext.cpp:23]: closeFD is 1 this looks scary

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp (hunk 5 above)
Comment (tomuben) [reply src/r-exasol/impl/rconnection/RReaderConnection.cpp]: Added shared/weak ptr.

File: src/connection.cpp [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        if (reader != nullptr) {
+            auto readerConnection = std::make_unique<exa::rconnection::RReaderConnection>(*reader);
+            retVal = readerConnection->create();
+            connectionContext->mConnection = std::move(readerConnection);
+        }
+    }
+    return retVal;
+}
+
+SEXP createWriteConnection(pRODBCHandle handle, SQLCHAR *query) {
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/connection.cpp:31]: Created a ConnectionContext class.

File: src/r-exasol/impl/rconnection/RReaderConnection.cpp (hunk 11 above)
Comment (tomuben) [reply src/r-exasol/impl/rconnection/RReaderConnection.cpp]: I added some documentation to the code now

File: src/r-exasol/impl/rconnection/RWriterConnection.h (hunk 12 above)
Comment (tomuben) [reply src/r-exasol/impl/rconnection/RWriterConnection.h]: I started to use namespace for that. Rconnection => ::Rconnection. I think that helps to understand better which type comes from where....

**Review by tkilias — CHANGES_REQUESTED** [review]
The c++ tests are quite hard to follow. Maybe explain what the tests do and what the particular steps in the test are.

File: tests/cpp/python_server_tests.py [hunk 15]
```diff
@@ -405,13 +439,14 @@ def writing_async_test_abort():
...
+    #writing_test()
+    reading_test_big()
+    #reading_http_test()
+    #con_controller_read_test()
+    #con_controller_echo_test()
+    #reading_async_test()
+    #reading_async_test_abort()
+    #writing_async_test()
+    #writing_async_test_abort()
+    #con_controller_read_test_with_error()
```
Comment (tkilias) [inline tests/cpp/python_server_tests.py:452]: tests are disabled

File: src/r_exasol/odbc/odbc_query_executor.h [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace exa {
+    /**
+     * Implements the functions used by @class AsyncExecutorImpl using the native ODBC functions.
+     * For the reader connection note:
+     *  The server will close the socket connection and finishes the the ODBC execution.
+     * For the writer connection note:
+     *  The client will close the socket connection when the data stream has finished -> This will trigger the server to finish the ODBC execution.
+     */
+    class OdbcQueryExecutor {
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/r_exasol/odbc/odbc_query_executor.h:19]: I have a added a new interface.

File: src/r_exasol/connection/connection_factory_impl.cpp [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+std::shared_ptr<exa::reader::Reader> exa::ConnectionFactoryImpl::createHttpReader(std::weak_ptr<Socket> socket) {
+    return std::make_shared<exa::reader::HttpChunkReader>(socket, getChunk());
+}
+
+std::shared_ptr<exa::writer::Writer> exa::ConnectionFactoryImpl::createHttpWriter(std::weak_ptr<Socket> socket) {
+    return std::make_shared<exa::writer::HttpChunkWriter>(socket, getChunk());
+}
+
+exa::Chunk & exa::ConnectionFactoryImpl::getChunk() {
+    static Chunk chunk;
```
[the remark this answers is not in the document]
Comment (tomuben) [reply src/r_exasol/connection/connection_factory_impl.cpp:19]: Done

File: src/r-exasol/ConnectionContext.cpp (hunk 13 above)
Comment (tomuben) [reply src/r-exasol/ConnectionContext.cpp]: I renamed the parameter name, it was quite confusing.

File: src/r_exasol/connection/connection_factory_impl.cpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+    return std::make_shared<exa::reader::HttpChunkReader>(socket, getChunk());
+}
+
+std::shared_ptr<exa::writer::Writer> exa::ConnectionFactoryImpl::createHttpWriter(std::weak_ptr<Socket> socket) {
+    return std::make_shared<exa::writer::HttpChunkWriter>(socket, getChunk());
+}
+
+exa::Chunk & exa::ConnectionFactoryImpl::getChunk() {
+    /**
+     * Chunk will be used accross different transfers. Reason is that it's allocation is quiete expansive (@see MAX_HTTP_CHUNK_SIZE)
```
Comment (tkilias) [inline src/r_exasol/connection/connection_factory_impl.cpp:20]:
```suggestion
     * Chunk will be used accross different transfers. Reason is that it's allocation is quiete expensive (@see MAX_HTTP_CHUNK_SIZE)
```

File: src/r_exasol/connection/connection_factory_impl.cpp [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+     * We use a static variable here, which is initialized the first time the function gets invoked,
+     * and then remains allocated until the program ends.
```
Comment (tkilias) [inline src/r_exasol/connection/connection_factory_impl.cpp:22]:
```suggestion
     * and then remains allocated until the program ends. We assume here, that only one read or write can happen at the same time.
```

File: doc/developer_guide/developer_guide.md [hunk 20]
```diff
@@ -45,5 +45,18 @@ When the client invokes ```readData()``` the library reads the server info (host
 Then the librart calls```asyncRODBCQueryStart()``` creates internally a new thread which sends the export statements via ODBC to the server (B1). The server then starts to write the data in CSV format to the formerly created TCP socket. Thus, the client receives the data from the server via the socket (not via ODBC).
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:45]:
```suggestion
Then the library calls```asyncRODBCQueryStart()``` creates internally a new thread which sends the export statements via ODBC to the server (B1). The server then starts to write the data in CSV format to the formerly created TCP socket. Thus, the client receives the data from the server via the socket (not via ODBC). 
```

File: doc/developer_guide/developer_guide.md [hunk 21]
```diff
@@ -45,5 +45,18 @@ When the client invokes ```readData()``` the library reads the server info (host
...
 
+### C level structure
+
+The following diagram describes the high level architecture of the C level package.
+![](./images/r-exasol-c_component.png)
+
+R uses the plain C interface, which delegates all call to class ```ConnectionContext```.
+When initiating a new connection, ```ConnectionContext``` instantiates ```ConnectionController```, which opens the socket
+and reads the meta-data.
+After receiving ```asuncRODBCQueryStart```, the ```ConnectionController``` triggers the asynchronous ODBC call (```ODBCAsyncExecutor```) and creates a ```HttpChunkReader``` or ```HttpChunkWriter``` instance, which is returned to the ```ConnectionContext``` class.
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:56]:
```suggestion
After receiving ```asyncRODBCQueryStart```, the ```ConnectionController``` triggers the asynchronous ODBC call (```ODBCAsyncExecutor```) and creates a ```HttpChunkReader``` or ```HttpChunkWriter``` instance, which is returned to the ```ConnectionContext``` class.
```

File: doc/developer_guide/developer_guide.md [hunk 22]
```diff
@@ -45,5 +45,18 @@ When the client invokes ```readData()``` the library reads the server info (host
...
+ConnectionContext then create a specific rconnection class: ```RWriterConnection``` or ```RReaderConnection```, and passes the reader/writer obtained from ```ConnectionController``` to it.
+Finally, ```ConnectionContext``` returns the ```SEXP``` instance (which encapsulates the rconnection data structure) to the R client, which can start reading/writing:
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:58]:
```suggestion
Finally, ```ConnectionContext``` returns the ```SEXP``` instance (which encapsulates the rconnection data structure) to the R client, which can start reading/writing.
```

File: doc/developer_guide/developer_guide.md [hunk 23]
```diff
@@ -45,5 +45,18 @@ When the client invokes ```readData()``` the library reads the server info (host
...
+
```
Comment (tkilias) [inline doc/developer_guide/developer_guide.md:59]: Maybe add a link to the code/documentation of the rconnection data structure and the RODBC stuff. Searching for it was difficult.

File: tests/scripts/README.md [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+They can also be used to run the tests locally.
+The test will automatically build a docker image for the specified R version and execute the 
+integration tests within the docker container.
+
+## Prerequisites
+
+* Docker
+* Bash
+
+##Available R version
```
Comment (tkilias) [inline tests/scripts/README.md:13]:
```suggestion
## Available R version
```

File: tests/scripts/README.md [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+First, you ned to select the R version.
```
Comment (tkilias) [inline tests/scripts/README.md:15]:
```suggestion
First, you need to select the R version.
```

**Review by tkilias — CHANGES_REQUESTED** [review]
see my comments

File: doc/developer_guide/developer_guide.md (hunk 23 above)
Comment (tomuben) [reply doc/developer_guide/developer_guide.md:59]: Good idea! Done.

File: tests/cpp/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,66 @@
+cmake_minimum_required(VERSION 3.19)
```
[the remark this answers is not in the document]
Comment (tomuben) [reply tests/cpp/CMakeLists.txt:1]: Indeed

**Review by tkilias — APPROVED** [review]
